### PR TITLE
Allow using slash command and other on cogs

### PR DESCRIPTION
**Pull request by noaione**
## Summary

This PR now should allow the usage of slash_command, user_command, and message_command from cogs. This implementation revert all the change that has been done and put it back where it should (especially the `cog.py` file).

In `ext/commands/cog` now there's new attributes which handle application (slash/user/message) in cogs. This should also fix error regarding `ctx` is not defined.

Currently, `SlashCommandGroup` is ignored in this PR so it might be broken.

Another thing I did:
- [x] Renamed `CheckFailure` in `app.errors` to `ApplicationCheckFailure` to set a difference with commands `CheckFailure` (can be reverted if you want)
- [x] Removed duplicates function and method in `app/commands`
- [x] Added some type-hinting for `app/commands`

Not tested yet:
- [ ] User command
- [ ] Message command

### Quick example
Example of this can be found here: https://github.com/noaione/pycord-cog-slash-example

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/utils.py [hunk 1]
```diff
@@ -448,14 +448,11 @@ def get(iterable: Iterable[T], **attrs: Any) -> Optional[T]:
             return elem
     return None
 
-async def get_or_fetch(obj, attr: str, id: int, *, default: Any = None):
+async def get_or_fetch(obj, attr: str, id: int):
```
Comment (Grace-codes) [inline discord/utils.py]: Why did you remove the default param? I think you should merge from slash once to sync the changes.

File: discord/ext/commands/errors.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 
 from typing import Optional, Any, TYPE_CHECKING, List, Callable, Type, Tuple, Union
 
-from discord.errors import ClientException, DiscordException
+from discord.errors import *
```
Comment (Grace-codes) [inline discord/ext/commands/errors.py:29]: Is a wildcard import needed here?

File: discord/ext/commands/core.py [hunk 3]
```diff
@@ -94,10 +93,7 @@
     'is_owner',
     'is_nsfw',
     'has_guild_permissions',
-    'bot_has_guild_permissions',
```
Comment (Grace-codes) [inline discord/ext/commands/core.py:97]: I'm assuming this was intended? So that people could use `commands.slash_command` etc.

File: discord/utils.py (hunk 1 above)
Comment (noaione) [reply discord/utils.py]: Already reverted, accidentally overwritten it.

File: discord/ext/commands/core.py (hunk 3 above)
Comment (noaione) [reply discord/ext/commands/core.py:97]: I think this was accidentally removed when I was reverting some changes

File: discord/ext/commands/errors.py (hunk 2 above)
Comment (noaione) [reply discord/ext/commands/errors.py:29]: I don't know if it's been removed, but originally you can access all the `discord.errors` from `commands.errors`.
I can revert this if needed

File: discord/ext/commands/core.py (hunk 3 above)
Comment (noaione) [reply discord/ext/commands/core.py:97]: Although, I think user might be better doing
`from discord.app import slash_command` it works too.

**Comment by noaione** [comment]
I created a quick example here: https://github.com/noaione/pycord-cog-slash-example

**Comment by BobDotCom** [comment]
We were planning on moving cogs to ext/cogs and possibly have separate classes for discord.Bot and commands.Bot.

**Comment by noaione** [comment]
Closing since I think this library already have their implementation ready(?).
